### DISCUSSION TITLE
Questions about Rendering Backends.

### DISCUSSION THREAD

**Discussion by jhimes144** [AvaloniaUI/Avalonia#17219] [Q&A]
Hi, I'm a bit confused about the current state on how Avalonia renders behind the scenes. Does Avalonia use skia exclusively all the time? When I choose a Metal backend for MacOS or Vulkan for Android, does Avalonia change the underlying render backend for skia (I'm assuming this is how it works)? It looks like skia supports all of the backends I've heard Avalonia using. Also, currently what is the default rendering backend for each platform, and does each platform support Vulkan currently? Thanks!

**Comment by kekekeks** [comment]
![image](https://github.com/user-attachments/assets/4d850471-c7fe-4905-9ef1-27a04d8d4afd)

No support for Vulkan on macOS, we don't ship MoltenVK and don't have integration with it.

**Reply by Kuumba123** [reply]
How do you setthe Backend to Direct3D11? I am pretty used to using WPF so when I heard about a cross platform version of WPF I knew I had to try it but I cant seem to pick  the backend. On Windows I want to use Direct3D11 but no matter what I do it always seems to just pick OpenGL. Any help would be apppreciated!

**Reply by lextm** [reply]
@Kuumba123 [The "cross platform version of WPF" is XPF, a closed source product from the same company](https://avaloniaui.net/xpf), not the so called Avalonia framework from this open source repo.

**Reply by Kuumba123** [reply]
I had no idea but it doesnt make sense that the same company that made Avalonia also made this XPF framework. I was refering to the git repro (Avalonia to be clear). Arlight back to the main issue. I am still un clear as to how I can  get Avalonia to use DirectX 11 as the backend for Windows (All forms of help would be nice) .

**Reply by thevortexcloud** [reply]
> I had no idea but it doesnt make sense that the same company that made Avalonia also made this XPF framework. 

XPF was created as a commercial solution so the Avalonia team could get some reliable revenue (as people are not donating huge amounts of money to Avalonia itself). It's targeted at people who have _extremely_ large code bases that use WPF and can't afford to rewrite all the presentation to work with Avalonia's version of XAML. Avalonia itself is **not** WPF, it's just heavily inspired by it. But a lot of WPF code needs to be rewritten (often minor edits) to work with Avalonia. By contrast, XPF **is** WPF (they literally forked the code provided by Microsoft), but uses Avalonia's rendering. So any code that works with WPF should also work with XPF with no changes.

>  I am still un clear as to how I can get Avalonia to use DirectX 11 as the backend for Windows (All forms of help would be nice) .

You need to set a platform specific option during boot.

https://api-docs.avaloniaui.net/docs/T_Avalonia_Win32PlatformOptions

There is an example here:

https://github.com/AvaloniaUI/Avalonia/blob/62597de97c8794f8c5a70da02596562328caade4/samples/GpuInterop/Program.cs#L18-L25

**Reply by Kuumba123** [reply]
That was what I was already doing. Is this some bug? I putted basicly the exact same code but when ever I get skia sharp to tell me the back end it always defaults to OpenGL: ```using Avalonia;
using Avalonia.Logging;
using Avalonia.Vulkan;
using System;

namespace AvaloniaTest
{
    internal class Program
    {
        // Initialization code. Don't use any Avalonia, third-party APIs or any
        // SynchronizationContext-reliant code before AppMain is called: things aren't initialized
        // yet and stuff might break.
        [STAThread]
        public static void Main(string[] args) => BuildAvaloniaApp()
            .StartWithClassicDesktopLifetime(args);

        // Avalonia configuration, don't remove; also used by visual designer.
        public static AppBuilder BuildAvaloniaApp()
        {
            return AppBuilder.Configure<App>()
                                        .UsePlatformDetect()
                .With(new Win32PlatformOptions
                {
                    RenderingMode = [Win32RenderingMode.AngleEgl],
                })
                .With(new X11PlatformOptions
                {
                    RenderingMode = [X11RenderingMode.Vulkan]
                })
                .With(new VulkanOptions
                {
                    VulkanInstanceCreationOptions = new VulkanInstanceCreationOptions
                    {
                        UseDebug = true
                    }
                })
                .LogToTrace(LogEventLevel.Debug, "Vulkan");
        }
    }
}
```
Here is how I am getting the info since for some reason my app crashes when I use RenderDoc
```
using Avalonia;
using Avalonia.Controls;
using Avalonia.Input;
using Avalonia.Media;
using Avalonia.Platform;
using Avalonia.Rendering.SceneGraph;
using Avalonia.Skia;
using Avalonia.Threading;
using SkiaSharp;
using System;

namespace AvaloniaTest
{
    public class SimpleSkiaControl : Control
    {
        public SimpleSkiaControl()
        {

        }
        protected override void OnPointerPressed(PointerPressedEventArgs e)
        {
            base.OnPointerPressed(e);
        }

        protected override void OnPointerReleased(PointerReleasedEventArgs e)
        {
            base.OnPointerReleased(e);
        }

        protected override void OnPointerMoved(PointerEventArgs e)
        {
            base.OnPointerMoved(e);
        }
        protected override void OnPointerEntered(PointerEventArgs e)
        {
            base.OnPointerEntered(e);
        }
        protected override void OnPointerExited(PointerEventArgs e)
        {
            base.OnPointerExited(e);
        }

        public override void Render(DrawingContext context)
        {
            base.Render(context);
            context.Custom(new SkiaDrawOp(Bounds));

            Dispatcher.UIThread.InvokeAsync(
                InvalidateVisual,
                DispatcherPriority.Background);
        }

        protected override Size MeasureOverride(Size availableSize)
        {
            return availableSize;
        }

        protected override Size ArrangeOverride(Size finalSize)
        {
            return finalSize;
        }

        private class SkiaDrawOp : ICustomDrawOperation
        {
            private readonly Rect _bounds;

            public SkiaDrawOp(Rect bounds)
            {
                _bounds = bounds;
            }

            public Rect Bounds => _bounds;
            public void Dispose() { }
            public bool Equals(ICustomDrawOperation? other) => false;
            public bool HitTest(Point p)
            {
                return _bounds.Contains(p);
            }


            public void Render(ImmediateDrawingContext context)
            {
                var leaseFeature = context.TryGetFeature<ISkiaSharpApiLeaseFeature>();
                if (leaseFeature == null)
                    return;

                using var lease = leaseFeature.Lease();
                var canvas = lease.SkCanvas;

                canvas.ClipRect(new SKRect(0,0,(float)_bounds.Width,(float)_bounds.Height));
                canvas.Clear(SKColors.Green);

                using var paint = new SKPaint
                {
                    Color = SKColors.Cyan,
                    StrokeWidth = 2,
                    IsAntialias = true,
                    Style = SKPaintStyle.Stroke
                };

                canvas.DrawCircle(150, 150, 80, paint);


                // Skia info
                var grContext = lease.GrContext; // Can be null if software rendering
                if (grContext != null)
                {
                    canvas.DrawText("Backend: " + grContext.Backend, 50, 50, paint);
                }
                else
                {
                    Console.WriteLine("Software rendering (no GRContext)");
                }

            }
        }
    }


}
```

**Comment by Kuumba123** [comment]
Thanks for the example. I should be able test it out and see if works sometime later this week. As for the whole XPF stuff yeah I total get and honestly thats probably a good thing (doing both paid and free software).

**Reply by Kuumba123** [reply]
my appolgies . I did some research on what this whole Angle API is and from what I know it is basicly just a openGL like API layer on top of directX11. A little strange that I have never heard of given the fact that I do know how to write basic shadders and I will be writting one for some of my other Avalonia programs!
